grpc/server/dronecore_server.cpp

```diff
@@ -93,169 +75,47 @@ class DroneCoreRPCImpl final : public DroneCoreRPC::Service
-                    }
-                    break;
-                case MissionState::ARMING:
-                    break;
-                case MissionState::ARMING_DONE:
-                    // TODO: There can be a race here if PX4 still listens to the armed flag in
-                    // the message DO_SET_MODE. Once it ignores it as in the spec, this is not
-                    // needed anymore.
-                    std::this_thread::sleep_for(std::chrono::seconds(2));
-                    device.mission().start_mission_async(
-                        std::bind(&receive_start_mission_result, _1));
-                    _mission_state = MissionState::STARTING;
-                    break;
-                case MissionState::STARTING:
-                    break;
-                case MissionState::STARTING_DONE:
-                    _mission_state = MissionState::MISSION;
-                    device.mission().subscribe_progress(
-                        std::bind(&receive_mission_progress, _1, _2));
-                    break;
-                case MissionState::MISSION_DONE:
-                    device.action().return_to_launch_async(
-                        std::bind(&receive_return_to_launch_result, _1));
-                    _mission_state = MissionState::RETURN;
-                    break;
-                case MissionState::RETURN:
-                    if (!device.telemetry().in_air()) {
-                        _mission_state = MissionState::RETURN_DONE;
-                    }
-                    break;
-                case MissionState::RETURN_DONE:
-                    device.action().disarm_async(std::bind(&receive_disarm_result, _1));
-                    break;
-                case MissionState::DONE:
-                    _finished = true;
-                    break;
-                case MissionState::ERROR:
-                    _finished = true;
-                    break;
-
-                case MissionState::MISSION:
-                // FALLTHROUGH
-                case MissionState::MISSION_PAUSING:
-                // FALLTHROUGH
-                case MissionState::PAUSE:
-                // FALLTHROUGH
-                case MissionState::MISSION_RESUMING:
-                // FALLTHROUGH
-                default:
-                    // not handled
-                    break;
-
-            }
-        }
-        if (_mission_state == MissionState::DONE) {
-            response->set_ret(true);
-        } else {
-            response->set_ret(false);
+        // TODO: there has to be a beter way than using std::future.
+        auto prom = std::make_shared<std::promise<void>>();
+        auto future_result = prom->get_future();
+
+        dc.device().mission().start_mission_async(
+        [prom, response](Mission::Result result) {
+            response->set_result(static_cast<dronecorerpc::MissionResult::Result>(result));
+            response->set_result_str(Mission::result_str(result));
```

Comment:
we are directly writing result to rpc client, I think we need to keep track of return status in server as well, in case if user calls other mission apis asynchronously

Reply:
How would these return stati mix up?